### PR TITLE
Bypass for CreateSession reqeust

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

- Fix the issue where too many request being prepared just waiting to be signed, while we need to make another CreateSession request from the same client to sign the requests.
- So, the requests will block the CreateSession to be made and hang

*Description of changes:*

- Bypass the throttle on client for CreateSession request, so that it will always get scheduled to be sent

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3express_credentials_provider.c [hunk 1]
```diff
@@ -506,6 +507,7 @@ static struct aws_s3express_session_creator *s_session_creator_new(
         /* Override endpoint only for tests. */
         .endpoint = impl->mock_test.endpoint_override ? impl->mock_test.endpoint_override : NULL,
         .user_data = session_creator,
+        .operation_name = operation_name,
```
Comment (waahm7) [inline source/s3express_credentials_provider.c:510]: trivial:
```suggestion
        .operation_name = aws_byte_cursor_from_c_str("CreateSession"),
```

File: source/s3_client.c [hunk 2]
```diff
@@ -1606,6 +1606,53 @@ static void s_s3_client_prepare_callback_queue_request(
     int error_code,
     void *user_data);
 
+static bool s_meta_request_create_more_requests(
```
Comment (waahm7) [inline source/s3_client.c:1609]: Naming suggestions:

- s_meta_request_can_create_more_requests
- s_should_client_create_more_requests
- ???

File: tests/s3_s3express_client_test.c [hunk 3]
```diff
@@ -256,14 +256,102 @@ static int s_s3express_client_put_test_real_server_helper(struct aws_allocator *
...
 
-TEST_CASE(s3express_client_put_test_small_real_server) {
+TEST_CASE(s3express_client_put_test_small) {
     (void)ctx;
-    return s_s3express_client_put_test_real_server_helper(allocator, MB_TO_BYTES(1));
+    return s_s3express_client_put_test_helper(allocator, MB_TO_BYTES(1));
 }
 
-TEST_CASE(s3express_client_put_test_large_real_server) {
+TEST_CASE(s3express_client_put_test_large) {
```
Comment (waahm7) [inline tests/s3_s3express_client_test.c:264]: Debatable: I wouldn't call 100MB large. Perhaps s3express_client_put_mpu_test would be better. Sometimes I specifically search for tests related to MPU / non-MPU, etc., and this naming would make it easier to find them.

File: source/s3_client.c (hunk 2 above)
Comment (DmitriyMusatkin) [inline source/s3_client.c:1609]: nit: s_meta_request_should_create_more_requests

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -294,11 +294,13 @@ if(ENABLE_MOCK_SERVER_TESTS)
...
-add_net_test_case(s3express_client_put_test_large_real_server)
-add_net_test_case(s3express_client_put_long_running_test_real_server)
-add_net_test_case(s3express_client_get_test_real_server)
+add_net_test_case(s3express_provider_long_running)
+add_net_test_case(s3express_client_put_test_small)
+add_net_test_case(s3express_client_put_test_large)
+add_net_test_case(s3express_client_put_test_multiple)
+add_net_test_case(s3express_client_put_long_running_test)
+add_net_test_case(s3express_client_get_test)
+add_net_test_case(s3express_client_get_test_multiple)
```
Comment (waahm7) [inline tests/CMakeLists.txt:303]: trivial / debatable: I prefer verbosity in test names, such as mentioning MPU/NON-MPU, the number of requests in multiples, and what is meant by a long-running test.

File: tests/s3_data_plane_tests.c [hunk 5]
```diff
@@ -1801,6 +1801,7 @@ static int s_test_s3_put_object_multiple_helper(struct aws_allocator *allocator,
             aws_string_destroy(filepath_str[i]);
         }
     }
+#undef NUM_REQUESTS
```
Comment (DmitriyMusatkin) [inline tests/s3_data_plane_tests.c:1804]: comment about this?

**Review by waahm7 — APPROVED** [review]
A bunch of trivial comments; feel free to merge and we can agonize over our names later, maybe.

File: tests/s3_s3express_client_test.c [hunk 6]
```diff
@@ -256,14 +256,102 @@ static int s_s3express_client_put_test_real_server_helper(struct aws_allocator *
...
     (void)ctx;
-    return s_s3express_client_put_test_real_server_helper(allocator, MB_TO_BYTES(100));
+    return s_s3express_client_put_test_helper(allocator, MB_TO_BYTES(100));
+}
+
+TEST_CASE(s3express_client_put_test_multiple) {
+    (void)ctx;
+
+#define NUM_REQUESTS 100
```
Comment (DmitriyMusatkin) [inline tests/s3_s3express_client_test.c:272]: cant this just be a local enum var as a work around for c99 not allowing const size_t to be size of array?

**Review by DmitriyMusatkin — APPROVED** [review]

File: source/s3_client.c [hunk 7]
```diff
@@ -1606,6 +1606,53 @@ static void s_s3_client_prepare_callback_queue_request(
...
+    struct aws_s3_meta_request *meta_request,
+    struct aws_s3_client *client,
+    uint32_t num_requests_in_flight,
+    const uint32_t max_requests_in_flight,
+    const uint32_t max_requests_prepare) {
+
+    /* CreateSession has high priority to bypass the checks. */
+    if (meta_request->type == AWS_S3_META_REQUEST_TYPE_DEFAULT) {
+        struct aws_s3_meta_request_default *meta_request_default = meta_request->impl;
+        if (aws_string_eq_c_str(meta_request_default->operation_name, "CreateSession")) {
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/s3_client.c:1619]: We don't have a way to specific the request type when we create the default type of meta request.

# TODO: maybe a reverse look up from the operation name to request type

**Review by graebm — APPROVED** [review]

**Pull request merged**